### PR TITLE
IP-1160 - Migrate to Python 3

### PR DESCRIPTION
**Pull request by kevinpetersavage**
* make code work in both python 2 and 3
* make generated code that works in both python 2 and 3
* fix tests in both python 2 and 3
* make setup able to produce a pip package that depends on either Avro or Avro-python3
* docker containers for python 2 and 3
* add ./build_and_run_tests_in_python_2_and_3.sh script

### CODE REVIEW

**Comment by greglever** [comment]
I'd like to reserve judgment until I can run it with the python 3 branches of the CIP API and validation service

File: protocols/util/factories/avro_factory.py [hunk 1]
```diff
@@ -107,7 +109,7 @@ def mock_complex_type(field, field_type, dependencies, version, fill_nullables):
     """
     value = None
     if field_type == 'record':
-        if isinstance(field.type, unicode):
+        if isinstance(field.type, str) or isinstance(field.type, basestring):
```
Comment (priesgo) [inline protocols/util/factories/avro_factory.py]: are we sure about `unicode` not needing to be here?

File: protocols_utils/code_generation/avrotools.py [hunk 2]
```diff
@@ -241,21 +242,21 @@ def handleBytes(self, datum):
             return self.sinkValue
 
     def handleInt(self, datum):
-        if not ((isinstance(datum, int) or isinstance(datum, long)) and
+        if not ((isinstance(datum, int)) and
```
Comment (priesgo) [inline protocols_utils/code_generation/avrotools.py]: are we sure `long` is not needed here?

File: protocols/util/factories/avro_factory.py (hunk 1 above)
Comment (kevinpetersavage) [reply protocols/util/factories/avro_factory.py]: Yes, unicode has been renamed string in python 3 but we use it in python 2. There is a basestring type from past.builtins that evaluates to the right thing dependent on version.

File: protocols_utils/code_generation/avrotools.py (hunk 2 above)
Comment (kevinpetersavage) [reply protocols_utils/code_generation/avrotools.py]: This is nearly right, but needs and additional import similar to the above. I will correct this.

**Comment by kevinpetersavage** [comment]
The `Parse` thing was a result of generated code checked in, but that if not generated by the current `builds.json`. Checking if we can remove these.

**Comment by priesgo** [comment]
some additional packages need to be removed. I'll give it a go

File: protocols_utils/code_generation/avrotools.py (hunk 2 above)
Comment (priesgo) [reply protocols_utils/code_generation/avrotools.py]: @kevinpetersavage not sure if this fixed, waiting your feedback before accepting

File: setup.py [hunk 3]
```diff
@@ -1,20 +1,30 @@
...
+interpreter_version = str(sys.version_info[0])
+target_version = enforced_version if enforced_version else interpreter_version
+if target_version == '2':
+    reqs += ["avro==1.7.7"]
+elif target_version == '3':
+    reqs += ["avro-python3==1.8.2"]
+else:
+    raise ValueError("Not supported python version {}".format(target_version))
+
 VERSION = "7.1.4"
```
Comment (priesgo) [inline setup.py]: version needs to be increased

**Pull request merged**